src/evo/cbtx.cpp

```diff
@@ -365,9 +365,11 @@ bool CheckCbTxBestChainlock(const CBlock& block, const CBlockIndex* pindex, cons
             return true;
         }
         uint256 curBlockCoinbaseCLBlockHash = pindex->GetAncestor(curBlockCoinbaseCLHeight)->GetBlockHash();
-        if (!chainlock_handler.VerifyChainLock(llmq::CChainLockSig(curBlockCoinbaseCLHeight, curBlockCoinbaseCLBlockHash, cbTx.bestCLSignature))) {
+        llmq::CChainLockSig cbtxcl = llmq::CChainLockSig(curBlockCoinbaseCLHeight, curBlockCoinbaseCLBlockHash, cbTx.bestCLSignature);
+        if (!chainlock_handler.VerifyChainLock(cbtxcl)) {
             return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-cbtx-invalid-clsig");
         }
+        chainlock_handler.ProcessNewChainLock(-1, cbtxcl, ::SerializeHash(cbtxcl));
```

Comment:
it should be not here, but when block is connecting.
There's still possibility that block would be refused, but chainlock should not be processed/submitted yet

Reply:
More over, even `CheckCbTxBestChainlock` can be called twice: once from `ProcessSpecialTxsInBlock` when flag `fJustCheck` is true - for simple pre-validation of block; and second time it's called from the same method `ProcessSpecialTxsInBlock` but when block is connected.

Reply:
Well; if the chain lock is valid; then the chain lock is valid and should be submitted?

Reply:
>There's still possibility that block would be refused, but chainlock should not be processed/submitted yet

It's a chainlock for some _previous_ block, not the one that is connected.

> More over, even `CheckCbTxBestChainlock` can be called twice: once from `ProcessSpecialTxsInBlock` when flag `fJustCheck` is true - for simple pre-validation of block; and second time it's called from the same method `ProcessSpecialTxsInBlock` but when block is connected.

`fJustCheck` is unrelated, it's `fCheckCbTxMerleRoots` that matters. And it is `true` in `ConnectBlock` and `false` in `RollforwardBlock` which is a completely different thing (not sure why it's `false` here tbh, feels like a bug... @PastaPastaPasta ).

Reply:
it's okay to have these checks as `false` in `RollforwardBlock` because there's intention to rollback block whatever it's valid or not. It is assumed that when it was connected it was valid -> to prevent bugs and for optimization ---> less checks.